src/clean.c

```diff
@@ -0,0 +1,394 @@
+ *   the Free Software Foundation, version 2 or later of the License.
+ *
+ *   This program is distributed in the hope that it will be useful,
+ *   but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *   GNU General Public License for more details.
+ *
+ *   You should have received a copy of the GNU General Public License
+ *   along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#define _GNU_SOURCE
+#include <ctype.h>
+#include <errno.h>
+#include <getopt.h>
+#include <sys/stat.h>
+#include <unistd.h>
+
+#include "swupd.h"
+
+/* Default value if --all is not passed. */
+const int DAYS_TO_KEEP_FILES = 3;
+
+static void print_help()
+{
+	fprintf(stderr,
+		"Usage:\n"
+		"	swupd clean [options]\n"
+		"\n"
+		"Remove cached data used for old updates from state directory.\n"
+		"\n"
+		"Options:\n"
+		"   --all                   Remove all the cached content not only old\n"
+		"   --dry-run               Just print files that would be removed\n"
+		"   -S, --statedir DIR      Specify alternate swupd state directory\n"
+		"   -h, --help              Display this help message\n"
+		"\n");
+}
+
+static struct {
+	int all;
+	int dry_run;
+} options;
+
+static struct {
+	int files_removed;
+} stats;
+
+static struct timespec now;
+
+static const struct option prog_opts[] = {
+	{ "help", no_argument, 0, 'h' },
+	{ "all", no_argument, &options.all, 1 },
+	{ "statedir", required_argument, 0, 'S' },
+	{ "dry-run", no_argument, &options.dry_run, 1 },
+	{ 0, 0, 0, 0 }
+};
+
+static bool parse_options(int argc, char **argv)
+{
+	int opt;
+	while ((opt = getopt_long(argc, argv, "hS:", prog_opts, NULL)) != -1) {
+		switch (opt) {
+		case '?':
+			return false;
+		case 'h':
+			print_help();
+			exit(0);
+		case 'S':
+			if (!optarg || !set_state_dir(optarg)) {
+				fprintf(stderr, "Invalid --statedir argument\n\n");
+				return false;
+			}
+			break;
+		case 0:
+			/* Handle options that don't have shortcut. */
+			break;
+		default:
+			fprintf(stderr, "Error: unrecognized option: -%c,\n\n", opt);
+			return false;
+		}
+	}
+	return true;
+}
+
+typedef bool(remove_predicate_func)(const char *dir, const struct dirent *entry);
+
+/* Remove files from path for which pred returns true.
+ * Currently it doesn't recursively remove directories.
+ *
+ * Unless --all is set, the files are also filtered by their age.
+ */
+static int remove_if(const char *path, remove_predicate_func pred)
+{
+	int ret = 0;
+	DIR *dir;
+	const int seconds = DAYS_TO_KEEP_FILES * 60 * 60 * 24;
+
+	dir = opendir(path);
+	if (!dir) {
+		return errno;
+	}
+
+	char *file = NULL;
+
+	while (true) {
+		free_string(&file);
+		ret = 0;
+
+		/* Reset errno to distinguish between a previous
+		 * failure and the end of stream. */
+		errno = 0;
+		struct dirent *entry;
+		entry = readdir(dir);
+		if (!entry) {
+			if (errno) {
+				ret = errno;
+			}
+			break;
```

Comment:
Optional. Since this is the only break, you could put the closedir() and return here.

Probably not worth it.